crypter.go

```diff
@@ -490,7 +490,9 @@ func (obj JSONWebEncryption) Decrypt(decryptionKey interface{}) ([]byte, error)
 	}
 
 	if plaintext == nil {
-		return nil, ErrCryptoFailure
+		if obj.ciphertext == nil || len(obj.ciphertext) > 0 {
```

Comment:
I don't fully understand this condition. Why is "plaintext is nil and also the ciphertext was nil" an error condition? It seems like if the ciphertext was nil, then the plaintext being nil is fairly expected.

Why isn't this simply
```go
if len(plaintext) == 0 && len(obj.ciphertext) != 0 {
```
?

Reply:
This approach leads to a nil pointer dereference in `TestJWENilProtected`:

```go
func TestJWENilProtected(t *testing.T) {
	key := []byte("1234567890123456")
	serialized := `{"unprotected":{"alg":"dir","enc":"A128GCM"}}`
	jwe, _ := ParseEncrypted(serialized)
	if _, err := jwe.Decrypt(key); err == nil {
		t.Error(err)
	}
}
```

In TestJWENilProtected, the JWE object has no protected header and no ciphertext, only an unprotected header. When the contents of this message are parsed `newBuffer()` sets obj.ciphertext to nil, thus calling len() on obj.ciphertext results in a nil pointer dereference because the ciphertext is set to nil by `newBuffer()`.

> It seems like if the ciphertext was nil, then the plaintext being nil is fairly expected.

When decrypting bytes representing an empty string, plaintext is nil, but ciphertext is not nil. Which means we should be able to accomplish the same check with `if plaintext == nil && ciphertext == nil`, but this check fails the following two cases in `TestVectorsJWECorrupt`:

```go
	msg, _ := ParseEncrypted(corruptCiphertext)
	_, err := msg.Decrypt(priv)
	if err != ErrCryptoFailure {
		t.Error("should detect corrupt ciphertext")
	}

	msg, _ = ParseEncrypted(corruptAuthtag)
	_, err = msg.Decrypt(priv)
	if err != ErrCryptoFailure {
		t.Error("should detect corrupt auth tag")
	}
```

If you think there's a more appropriate way to handle this, please let me know.

Reply:
> When the contents of this message are parsed newBuffer() sets obj.ciphertext to nil, thus calling len() on obj.ciphertext results in a nil pointer dereference.

Calling len on a nil slice shouldn't result in a nil pointer dereference, it should result in the integer zero: https://go.dev/play/p/nC1e1TpD9U- . Even calling `len(nil)` directly doesn't result in a nil pointer dereference, it results in a [compile-time error](https://go.dev/play/p/V23cBGs1SpQ).

When I make this change to crypter.go, I do see a nil pointer dereference inside TestJWENilProtected, but it's not from len(obj.ciphertext):
```
❯ go test ./...
--- FAIL: TestJWENilProtected (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x718cb3]

goroutine 57 [running]:
testing.tRunner.func1.2({0x7946a0, 0xab15f0})
	/usr/local/go/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1529 +0x39f
panic({0x7946a0, 0xab15f0})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/go-jose/go-jose/v3.JSONWebEncryption.Decrypt({{{0x0, 0x0}, 0x0, {0xc00051f8bb, 0x3}, {0x0, 0x0}, {0x0, 0x0, 0x0}, ...}, ...}, ...)
	/home/aaron/Documents/github.com/go-jose/go-jose/crypter.go:497 +0x453
github.com/go-jose/go-jose/v3.TestJWENilProtected(0xc0005031e0)
	/home/aaron/Documents/github.com/go-jose/go-jose/jwe_test.go:332 +0x10e
testing.tRunner(0xc0005031e0, 0x831bd0)
	/usr/local/go/src/testing/testing.go:1576 +0x10b
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1629 +0x3ea
FAIL	github.com/go-jose/go-jose/v3	7.385s
```

This failure is on line 497, where crypter tries to access a method on obj.protected:
```go
	if comp := obj.protected.getCompression(); comp != "" {
```

Presumably, previously the function would error out at the "if plaintext == nil" check when obj.protected was nil, because decryption would have failed if there was no protected data. With the addition of "&& len(obj.ciphertext) == 0", it's now able to get past that check and run into this nil pointer dereference instead.

> When decrypting bytes representing an empty string, plaintext is nil, but ciphertext is not nil.

Right, and that's totally fine. What I'm talking about is the case where _ciphertext_ is nil, in which case plaintext _must_ also be nil: there cannot be a 0-length cipertext which decrypts to a non-zero-length plaintext.

Reply:
Ah, yeah I see where I went wrong here. I was under the impression that ciphertext was *byteBuffer (which also wouldn't work, as you could only `len(obj.ciphertex.bytes())`)

I'll take another look and revise this today.